1. Выполнить токенизацию текста любимой книги с использованием токенизаторов библиотек nltk и spacy. Описать отличия, достоинства и недостатки токенизаторов.
2. Определить лексическое разнообразие книги.
3. Написать собственный токенизатор, выполненный в качестве метода класса.

In [1]:
import nltk
import nltk.tokenize

import timeit

In [2]:
#загружаем все 122 главы в одну переменную
text = ''
for i in range(1, 123):
    path = 'C:\\Users\\user\\Desktop\\DeepAutomation\\Tokenizing\\hw1\\chapters\\' + str(i).rjust(4, '0') + '.txt'
    with open(path) as f:
        text += ' '.join(f.readlines())
len(text)

3752789

In [6]:
start_time = timeit.default_timer()

whitespace = nltk.tokenize.WhitespaceTokenizer()
tokens_whitespace = whitespace.tokenize(text)

print(timeit.default_timer() - start_time)

0.11143980000633746


In [7]:
tokens_whitespace

['Chapter',
 '1:',
 'A',
 'Day',
 'of',
 'Very',
 'Low',
 'Probability',
 'Disclaimer:',
 'J.',
 'K.',
 'Rowling',
 'owns',
 'Harry',
 'Potter,',
 'and',
 'no',
 'one',
 'owns',
 'the',
 'methods',
 'of',
 'rationality.',
 'This',
 'fic',
 'is',
 'widely',
 'considered',
 'to',
 'have',
 'really',
 'hit',
 'its',
 'stride',
 'starting',
 'at',
 'around',
 'Chapter',
 '5.',
 'If',
 'you',
 'still',
 "don't",
 'like',
 'it',
 'after',
 'Chapter',
 '10,',
 'give',
 'up.',
 'Please',
 'visit',
 'HPMOR',
 'DOT',
 'COM',
 'for:',
 '*',
 'Easy',
 'email',
 'notification',
 'system,',
 'RSS',
 'feed,',
 'and',
 'Twitter',
 'feed',
 'for',
 'new',
 'chapters;',
 '*',
 'Current',
 "Author's",
 'Notes',
 'and',
 'progress',
 'updates;',
 '*',
 'Lovely',
 'fan-made',
 'book-style',
 'PDF',
 'version;',
 '*',
 'Ad-free',
 'mirror',
 'of',
 'the',
 'text;',
 '*',
 'ePUB',
 'and',
 'MOBI',
 'e-texts;',
 '*',
 'Ongoing',
 'podcast',
 'of',
 'the',
 'story;',
 '*',
 'Fan',
 'art',
 'in',
 'vast',
 'qua

In [24]:
nltk.tokenize.word_tokenize(text)

['Chapter',
 '1',
 ':',
 'A',
 'Day',
 'of',
 'Very',
 'Low',
 'Probability',
 'Disclaimer',
 ':',
 'J.',
 'K.',
 'Rowling',
 'owns',
 'Harry',
 'Potter',
 ',',
 'and',
 'no',
 'one',
 'owns',
 'the',
 'methods',
 'of',
 'rationality',
 '.',
 'This',
 'fic',
 'is',
 'widely',
 'considered',
 'to',
 'have',
 'really',
 'hit',
 'its',
 'stride',
 'starting',
 'at',
 'around',
 'Chapter',
 '5',
 '.',
 'If',
 'you',
 'still',
 'do',
 "n't",
 'like',
 'it',
 'after',
 'Chapter',
 '10',
 ',',
 'give',
 'up',
 '.',
 'Please',
 'visit',
 'HPMOR',
 'DOT',
 'COM',
 'for',
 ':',
 '*',
 'Easy',
 'email',
 'notification',
 'system',
 ',',
 'RSS',
 'feed',
 ',',
 'and',
 'Twitter',
 'feed',
 'for',
 'new',
 'chapters',
 ';',
 '*',
 'Current',
 'Author',
 "'s",
 'Notes',
 'and',
 'progress',
 'updates',
 ';',
 '*',
 'Lovely',
 'fan-made',
 'book-style',
 'PDF',
 'version',
 ';',
 '*',
 'Ad-free',
 'mirror',
 'of',
 'the',
 'text',
 ';',
 '*',
 'ePUB',
 'and',
 'MOBI',
 'e-texts',
 ';',
 '*',
 'Ongoin

In [9]:
import spacy

In [21]:
#spacy ругается на слишком длинный текст, поэтому прогоню частично
len(text) # = 3 752 789

text = text[:999999]
len(text)

999999

In [22]:
start_time = timeit.default_timer()

model = spacy.load('en_core_web_sm')
doc = model(text)
tokens = [token.text for token in doc]
print(timeit.default_timer() - start_time)

17.849273399973754


In [12]:
tokens

['Chapter',
 '1',
 ':',
 'A',
 'Day',
 'of',
 'Very',
 'Low',
 '\n ',
 'Probability',
 '\n \n ',
 'Disclaimer',
 ':',
 'J.',
 'K.',
 'Rowling',
 'owns',
 'Harry',
 'Potter',
 ',',
 'and',
 'no',
 'one',
 'owns',
 'the',
 '\n ',
 'methods',
 'of',
 'rationality',
 '.',
 '\n ',
 'This',
 'fic',
 'is',
 'widely',
 'considered',
 'to',
 'have',
 'really',
 'hit',
 'its',
 'stride',
 '\n ',
 'starting',
 'at',
 'around',
 'Chapter',
 '5',
 '.',
 'If',
 'you',
 'still',
 'do',
 "n't",
 'like',
 'it',
 'after',
 '\n ',
 'Chapter',
 '10',
 ',',
 'give',
 'up',
 '.',
 '\n ',
 'Please',
 'visit',
 'HPMOR',
 'DOT',
 'COM',
 'for',
 ':',
 '\n ',
 '*',
 'Easy',
 'email',
 'notification',
 'system',
 ',',
 'RSS',
 'feed',
 ',',
 'and',
 'Twitter',
 'feed',
 'for',
 '\n ',
 'new',
 'chapters',
 ';',
 '\n ',
 '*',
 'Current',
 'Author',
 "'s",
 'Notes',
 'and',
 'progress',
 'updates',
 ';',
 '\n ',
 '*',
 'Lovely',
 'fan',
 '-',
 'made',
 'book',
 '-',
 'style',
 'PDF',
 'version',
 ';',
 '\n ',
 '*'

считает спецсимволы типа "\n" за токены, попробую убрать их

In [14]:
text = text.replace('\n', '')

In [25]:
start_time = timeit.default_timer()

model = spacy.load('en_core_web_sm')
doc = model(text)
tokens = [token.text for token in doc]
print(timeit.default_timer() - start_time)

17.92221220000647


In [16]:
tokens

['Chapter',
 '1',
 ':',
 'A',
 'Day',
 'of',
 'Very',
 'Low',
 'Probability',
 ' ',
 'Disclaimer',
 ':',
 'J.',
 'K.',
 'Rowling',
 'owns',
 'Harry',
 'Potter',
 ',',
 'and',
 'no',
 'one',
 'owns',
 'the',
 'methods',
 'of',
 'rationality',
 '.',
 'This',
 'fic',
 'is',
 'widely',
 'considered',
 'to',
 'have',
 'really',
 'hit',
 'its',
 'stride',
 'starting',
 'at',
 'around',
 'Chapter',
 '5',
 '.',
 'If',
 'you',
 'still',
 'do',
 "n't",
 'like',
 'it',
 'after',
 'Chapter',
 '10',
 ',',
 'give',
 'up',
 '.',
 'Please',
 'visit',
 'HPMOR',
 'DOT',
 'COM',
 'for',
 ':',
 '*',
 'Easy',
 'email',
 'notification',
 'system',
 ',',
 'RSS',
 'feed',
 ',',
 'and',
 'Twitter',
 'feed',
 'for',
 'new',
 'chapters',
 ';',
 '*',
 'Current',
 'Author',
 "'s",
 'Notes',
 'and',
 'progress',
 'updates',
 ';',
 '*',
 'Lovely',
 'fan',
 '-',
 'made',
 'book',
 '-',
 'style',
 'PDF',
 'version',
 ';',
 '*',
 'Ad',
 '-',
 'free',
 'mirror',
 'of',
 'the',
 'text',
 ';',
 '*',
 'ePUB',
 'and',
 'MOB

Время составляет около ~17.9 секунд против доль секунды nltk

In [10]:
nltk.concordance()

AttributeError: module 'nltk' has no attribute 'concordance'